### PR TITLE
Fix: feature image virtual media

### PR DESCRIPTION
**Pull request by KMchaudhary**
**Fixes:** https://github.com/rtCamp/godam/issues/1977

This pull request introduces robust support for using GoDAM media items as featured images in both the block and classic editors. The main changes ensure that placeholder GoDAM IDs are properly handled and swapped for real WordPress attachment IDs once the media entry is created, preventing issues where invalid IDs could be saved or displayed. The implementation includes new logic for the classic editor, improved handling in the block editor, and comprehensive unit tests.

**Featured image handling improvements:**

* Added `classic-featured-image.js` to intercept and defer setting GoDAM IDs as featured images in the classic editor, only applying the real attachment ID once it exists, and preventing invalid IDs from being saved or displayed.
* Updated `godam-media-frame-shared.js` to invoke the new classic editor logic (`setupClassicFeaturedImage` and `resolveClassicFeaturedImage`) during GoDAM tab activation and after attachment creation. [[1]](diffhunk://#diff-a7878e67888eef229132a39357ea6378cbaf2c7cdc41b7b4bc69f4333e5fd9a4R19) [[2]](diffhunk://#diff-a7878e67888eef229132a39357ea6378cbaf2c7cdc41b7b4bc69f4333e5fd9a4R203-R205) [[3]](diffhunk://#diff-a7878e67888eef229132a39357ea6378cbaf2c7cdc41b7b4bc69f4333e5fd9a4R313-R318)

**Block editor and post entity integration:**

* Replaced the logic for updating the featured image in the block editor: added `replaceVirtualIdInFeaturedImage()` to swap the GoDAM placeholder ID for the real attachment ID on the post entity, ensuring the correct image is displayed and saved.

**Testing and reliability:**

* Added `classic-featured-image.test.js` with comprehensive tests for the new classic editor featured image deferral and resolution logic, covering edge cases and ensuring correct integration.

**Codebase simplification:**

* Removed the now-unnecessary `checkIfFeatureImage()` logic and related conditional checks, as the new approach directly handles both block and classic editor contexts. [[1]](diffhunk://#diff-a7878e67888eef229132a39357ea6378cbaf2c7cdc41b7b4bc69f4333e5fd9a4L107-R145) [[2]](diffhunk://#diff-a7878e67888eef229132a39357ea6378cbaf2c7cdc41b7b4bc69f4333e5fd9a4L153-R171)

## Demo

https://github.com/user-attachments/assets/b115b0a0-8b1d-482a-9ec8-8d0133af5797

### CODE REVIEW

File: assets/src/js/media-library/views/godam-media-frame-shared.js [hunk 1]
```diff
@@ -292,6 +310,12 @@ const GoDAMMediaFrameShared = {
 				// so that inner blocks are resolved without relying on React effect timing.
 				replaceVirtualIdInCoreImageBlocks( data.id, attachment.id );
 
+				// The featured image lives on the post entity rather than on block
+				// attributes, so it needs its own swap. The classic editor gets no
+				// placeholder at all — its pick was parked, and is released here.
+				replaceVirtualIdInFeaturedImage( data.id, attachment.id );
+				resolveClassicFeaturedImage( data.id, attachment.id );
```
Comment (subodhr258) [inline assets/src/js/media-library/views/godam-media-frame-shared.js:341]: **Failed `create-media-entry` silently drops the classic featured-image pick.** `resolveClassicFeaturedImage()` only runs inside the `if ( response.success )` block. If the request fails or returns non-success, the outer `catch` at the bottom of `processGoDAMItem` swallows it, the parked id is never released, and `deferredVirtualId` stays stuck. The meta box keeps its previous value with no error surfaced, so the user believes a featured image was set when nothing changed. Consider releasing the park (or clearing it with a visible fallback) on the failure path too.

File: assets/src/js/media-library/views/godam-media-frame-shared.js [hunk 2]
```diff
@@ -104,29 +105,44 @@ function replaceVirtualIdInCoreImageBlocks( virtualMediaId, realId ) {
...
-};
+	if ( currentFeaturedMedia === undefined || currentFeaturedMedia === null ) {
+		return;
+	}
+
+	if ( String( currentFeaturedMedia ) !== String( virtualMediaId ) ) {
+		return;
+	}
+
+	dispatch( 'core/editor' ).editPost( { featured_media: realId } );
```
Comment (subodhr258) [inline assets/src/js/media-library/views/godam-media-frame-shared.js]: **Guard this `editPost` the way the read above it is guarded.** The `getEditedPostAttribute` read on line 131 sits in a try/catch, but this `dispatch( 'core/editor' ).editPost()` does not. A throw here unwinds into `processGoDAMItem`'s swallow-all `catch`, which then skips `resolveClassicFeaturedImage()` and both the `godam-virtual-attachment-created` and `godam-attachment-browser:changed` dispatches for this item. `replaceVirtualIdInCoreImageBlocks` is self-guarded, so this is the one exposed call in that block. Wrapping it in the same defensive guard keeps a bad dispatch from taking out the events other code depends on.

File: assets/src/js/media-library/classic-featured-image.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	}
+
+	// Kept for resolveClassicFeaturedImage(): it runs while the modal still reports
+	// 'godam' as its content mode, so releasing through the wrapper would park the real
+	// ID too and the pick would never land.
+	featuredImage.godamOriginalSet = originalSet;
+
+	featuredImage.set = function( id ) {
+		if ( isDeferrableGoDAMId( id ) ) {
+			deferredVirtualId = id;
```
Comment (subodhr258) [inline assets/src/js/media-library/classic-featured-image.js:162]: **Save-before-resolve keeps the old featured image (classic editor).** Parking suppresses the original `set()`, so the meta box `_thumbnail_id` hidden field still holds the previous value until `create-media-entry` answers and `resolveClassicFeaturedImage()` releases the real id. If the user clicks Update/Publish inside that sub-second window, the save persists the stale id and the newly chosen image is lost until they save again. Worth confirming that timing is acceptable, or showing a brief in-flight indicator while a pick is being resolved.

File: assets/src/js/media-library/classic-featured-image.test.js [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+			frame,
+			featuredImage: { set },
+		},
+	};
+
+	return { set, frame };
+};
+
+describe( 'classic editor featured image deferral', () => {
+	afterEach( () => {
```
Comment (subodhr258) [inline assets/src/js/media-library/classic-featured-image.test.js:60]: **`deferredVirtualId` is module state that `afterEach` never resets.** It only deletes `window.wp` and restores mocks, so a test that parks a pick (e.g. 'parks a GoDAM pick') leaves `deferredVirtualId` set for the next test. That makes the suite order-dependent and able to pass or fail spuriously when tests are reordered or new ones are inserted before the existing ones. Drain the park in `afterEach` (call `resolveClassicFeaturedImage` to clear it, or export a small reset for tests).

File: assets/src/js/media-library/views/godam-media-frame-shared.js [hunk 5]
```diff
@@ -185,6 +200,9 @@ const GoDAMMediaFrameShared = {
 	GoDAMCreate() {
 		const state = this.state();
 
+		// Hold back a classic-editor featured image pick until its attachment exists.
+		setupClassicFeaturedImage();
```
Comment (subodhr258) [inline assets/src/js/media-library/views/godam-media-frame-shared.js]: **Nit (altitude): this installs a global core patch on every GoDAM tab render, not just featured-image contexts.** `setupClassicFeaturedImage()` permanently wraps the core global `wp.media.featuredImage.set` the first time any GoDAM tab renders, including the Insert-Media and gallery flows where no featured image is ever picked. It is idempotent and transparent for non-deferrable ids, so not a live bug, but it widens the blast radius of a core-global override past the context the fix targets. Consider only wrapping when the frame is actually a featured-image frame.

File: assets/src/js/media-library/classic-featured-image.js [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+function isDeferrableGoDAMId( id ) {
+	const frame = window.wp?.media?.frame;
+
+	if ( 'godam' !== frame?.content?.mode?.() ) {
+		return false;
+	}
+
+	const selected = frame.state?.()?.get?.( 'selection' )?.single?.();
+
+	return !! selected && String( selected.id ) === String( id );
```
Comment (subodhr258) [inline assets/src/js/media-library/classic-featured-image.js]: **Nit (reuse): the `String( a ) === String( b )` id-equality idiom is repeated** across `isDeferrableGoDAMId`, `resolveClassicFeaturedImage`, `replaceVirtualIdInFeaturedImage`, and `replaceVirtualIdInCoreImageBlocks`. A shared `isSameId( a, b )` helper in the media-library utility module would keep the virtual-vs-real id comparison in one place if the matching rule ever needs to change (trimming, coercion, etc.).

**Review by subodhr258 — COMMENTED** [review]
Review notes on the featured-image virtual-media fix. The core-frame wiring, `wp.media.frame` reset path, and the sync-editPost / async-create ordering all check out. A few things worth a look, correctness first:

**Comment by subodhr258** [comment]
### Re-review

Went through the diff again end to end, and ran the new suite plus lint on the branch.

- `classic-featured-image.test.js`: 15/15 pass
- ESLint on the four changed JS files: clean

The core design holds up well. Parking the classic `set()` and releasing the real ID after `create-media-entry` answers is the right call, the block vs classic paths stay independent (the classic wrap only installs on a `featured-image` frame, and the block swap edits the post entity), and every path is keyed by ID so multi-select and slow failures do not cross-talk. Failure and clear paths are covered.

A few minor notes, all low severity, none blocking:

**1. Two quick classic picks can strand the first and leave the placeholder stuck**
`deferredVirtualId` and `metaBoxSnapshot` are single slots, so they can only track one pick at a time. In the classic editor: pick A as featured image, then reopen and pick B before A's `create-media-entry` resolves. `showInFlightPlaceholder()` then overwrites the snapshot with the "Setting featured image…" markup, and `deferredVirtualId` moves to B. When A resolves it no-ops on the ID mismatch (so A is never applied), and if B then fails, `restoreMetaBox()` puts the "Setting featured image…" text back as the "previous" image, leaving the meta box stuck on the placeholder until reload. Narrow race given the modal is single-select and closes on select, but worth a comment at minimum. No test covers a second pick landing while one is parked.

**2. Stuck placeholder if core's `set()` gets `'0'` back**
In `resolveClassicFeaturedImage()`, `metaBoxSnapshot` is nulled right before delegating to core's `set(realId)`. If `get-post-thumbnail-html` returns `'0'`, core does not repaint `#postimagediv .inside`, so the placeholder is left with no restore path. Very edge for a freshly created attachment, listing it for completeness.

**3. `isSameId` is not reused in `core-media-blocks-extension.js:56`**
`ids.js` says it exists to keep the virtual/real match rule "in a single place", but the identical `String(a) === String(b)` at `core-media-blocks-extension.js:56` was left unconverted, so the rule now lives in two spots. That file is outside this PR's touch set, so this is optional. If converted it needs the existing null/undefined allowance kept, e.g. `mediaId == null || isSameId( mediaId, virtualMediaId )`, since `isSameId` returns false for null/undefined.

Nothing here blocks merge.

**Review by subodhr258 — APPROVED** [review]
